In [48]:
import pandas as pd

In [49]:
# Make a reference to the books.csv file path
csv_path1 = "Resources/students_complete.csv"
csv_path2 = "Resources/schools_complete.csv"


In [50]:
# Import the students_complete.csv file as a DataFrame
df1 = pd.read_csv(csv_path1, encoding="utf-8")

In [51]:
# Import the schools.csv file as a DataFrame
df2 = pd.read_csv(csv_path2, encoding="utf-8")

In [52]:
df1.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  
0             66          79  
1             94          61  
2             90          60  
3             67          58  
4             97          84

In [53]:
df2.head()

School ID            school_name      type  size   budget
0          0      Huang High School  District  2917  1910635
1          1   Figueroa High School  District  2949  1884411
2          2    Shelton High School   Charter  1761  1056600
3          3  Hernandez High School  District  4635  3022020
4          4    Griffin High School   Charter  1468   917500

# District Summary


In [54]:
#Merging two csv files
merge_df = pd.merge(df2, df1, on="school_name")
merge_df


School ID         school_name      type  size   budget  Student ID  \
0              0   Huang High School  District  2917  1910635           0   
1              0   Huang High School  District  2917  1910635           1   
2              0   Huang High School  District  2917  1910635           2   
3              0   Huang High School  District  2917  1910635           3   
4              0   Huang High School  District  2917  1910635           4   
...          ...                 ...       ...   ...      ...         ...   
39165         14  Thomas High School   Charter  1635  1043130       39165   
39166         14  Thomas High School   Charter  1635  1043130       39166   
39167         14  Thomas High School   Charter  1635  1043130       39167   
39168         14  Thomas High School   Charter  1635  1043130       39168   
39169         14  Thomas High School   Charter  1635  1043130       39169   

            student_name gender grade  reading_score  math_score  
0           Paul Bradley      M   9th             66          79  
1           Victor Smith      M  12th             94          61  
2        Kevin Rodriguez      M  12th             90          60  
3      Dr. Richard Scott      M  12th             67          58  
4             Bonnie Ray      F   9th             97          84  
...                  ...    ...   ...            ...         ...  
39165       Donna Howard      F  12th             99          90  
39166          Dawn Bell      F  10th             95          70  
39167     Rebecca Tanner      F   9th             73          84  
39168       Desiree Kidd      F  10th             99          90  
39169    Carolyn Jackson      F  11th             95          75  

[39170 rows x 11 columns]

In [55]:
#Looking for missing values
merge_df.count()

School ID        39170
school_name      39170
type             39170
size             39170
budget           39170
Student ID       39170
student_name     39170
gender           39170
grade            39170
reading_score    39170
math_score       39170
dtype: int64

In [56]:
#Calculating total number of schools
count_of_schools= len(merge_df["School ID"].unique())
count_of_schools

15

In [57]:
#Calculating total number of students
merge_df["student_name"].value_counts()

Michael Smith       26
David Johnson       16
Michael Williams    16
David Smith         15
Daniel Johnson      14
                    ..
Austin Davis         1
Taylor Avery         1
Shelly Davis         1
Rachel Moss          1
James Clark II       1
Name: student_name, Length: 32715, dtype: int64

In [58]:
#Calculating total number of students
count_of_students=len(merge_df)
count_of_students

39170

In [59]:
#Calculating total budget
total_budget=df2["budget"].sum()
total_budget


24649428

In [60]:
#Finding average math score
aver_math_score=merge_df["math_score"].mean()
aver_math_score

78.98537145774827

In [61]:
#Finding average reading score
aver_reading_score=merge_df["reading_score"].mean()
aver_reading_score

81.87784018381414

In [62]:
#Passing Math (The percentage of students that passed math.)
amount_passed_math = len(merge_df.loc[merge_df["math_score"] >=70])
amount_passed_math


29370

In [63]:
perc_of_students_passing_math=(amount_passed_math/count_of_students)*100
perc_of_students_passing_math

74.9808526933878

In [64]:
#Passing Reading (The percentage of students that passed reading.)
amount_passed_reading = len(merge_df.loc[merge_df["reading_score"] >=70])
amount_passed_reading

33610

In [65]:
perc_of_students_passing_reading=(amount_passed_reading/count_of_students)*100
perc_of_students_passing_reading

85.80546336482001

In [66]:
#Overall Passing (The percentage of students that passed math and reading.)
amount_passed_math_and_reading=merge_df.loc[(merge_df["math_score"] >=70) & (
    merge_df["reading_score"] >=70), :]
len(amount_passed_math_and_reading)
                        

25528

In [67]:
perc_of_students_passing_both=(len(amount_passed_math_and_reading)/count_of_students)*100
perc_of_students_passing_both

65.17232575950983

In [103]:
# Create a new table consolodating above calculations
district_summary_0 = pd.DataFrame({"Total Number of Schools": [count_of_schools],
                                   "Total Number of Students": [count_of_students],
                                   "Total Budget": [total_budget],
                                   "Avg. Math Score": [aver_math_score],
                                   "Avg. Reading Score": [aver_reading_score],
                                   "% Passing Math": [perc_of_students_passing_math],
                                   "% Passing Reading":[perc_of_students_passing_reading],
                                   "% Passing Both": [perc_of_students_passing_both]
                                   })
district_summary = district_summary_0.round(2)
#district_summary['Total Budget'] = ['${:,.2f}'.format(x) for x in district_summary['Total Budget']]

district_summary['Total Budget'] = district_summary['Total Budget'].map("${:,.2f}".format)
district_summary

Total Number of Schools  Total Number of Students    Total Budget  \
0                       15                     39170  $24,649,428.00   

   Avg. Math Score  Avg. Reading Score  % Passing Math  % Passing Reading  \
0            78.99               81.88           74.98              85.81   

   % Passing Both  
0           65.17

# School Summary
Create an overview table that summarizes key metrics about each school, including:
School Name
School Type
Total Students
Total School Budget
Per Student Budget
Average Math Score
Average Reading Score
% Passing Math (The percentage of students that passed math.)
% Passing Reading (The percentage of students that passed reading.)
% Overall Passing (The percentage of students that passed math and reading.)


In [69]:
#Finding school budget for each school
#schools=merge_df.groupby(["school_name"])
d1 = df2.loc[:,[ "school_name",'type', 'size', 'budget']]

schools = pd.DataFrame(d1)

schools.rename(columns={'type': 'School Type',
                        'school_name':'School Name',
                        'size': 'Total Students',
                        'budget': 'Total School Budget', 
                        }, inplace=True)

schools['Per Student Budget'] = schools['Total School Budget']/schools['Total Students']

schools.head()

School Name School Type  Total Students  Total School Budget  \
0      Huang High School    District            2917              1910635   
1   Figueroa High School    District            2949              1884411   
2    Shelton High School     Charter            1761              1056600   
3  Hernandez High School    District            4635              3022020   
4    Griffin High School     Charter            1468               917500   

   Per Student Budget  
0               655.0  
1               639.0  
2               600.0  
3               652.0  
4               625.0

In [70]:
#Calculating and listing average test score for each school
d2 = merge_df.groupby(['school_name'])
score = pd.DataFrame(round(d2.mean(),2))   
score1 = score.loc[:,['reading_score','math_score']]


score1=score1.reset_index()
score1.rename(columns={'school_name':'School Name',
                      'reading_score':'Aver. Reading Score',
                      'math_score': 'Aver. Math Score',
                     }, inplace=True)
score1

School Name  Aver. Reading Score  Aver. Math Score
0      Bailey High School                81.03             77.05
1     Cabrera High School                83.98             83.06
2    Figueroa High School                81.16             76.71
3        Ford High School                80.75             77.10
4     Griffin High School                83.82             83.35
5   Hernandez High School                80.93             77.29
6      Holden High School                83.81             83.80
7       Huang High School                81.18             76.63
8     Johnson High School                80.97             77.07
9        Pena High School                84.04             83.84
10  Rodriguez High School                80.74             76.84
11    Shelton High School                83.73             83.36
12     Thomas High School                83.85             83.42
13     Wilson High School                83.99             83.27
14     Wright High School                83.96             83.68

In [71]:
add= pd.merge(schools,score1,how='left', on='School Name')
add

School Name School Type  Total Students  Total School Budget  \
0       Huang High School    District            2917              1910635   
1    Figueroa High School    District            2949              1884411   
2     Shelton High School     Charter            1761              1056600   
3   Hernandez High School    District            4635              3022020   
4     Griffin High School     Charter            1468               917500   
5      Wilson High School     Charter            2283              1319574   
6     Cabrera High School     Charter            1858              1081356   
7      Bailey High School    District            4976              3124928   
8      Holden High School     Charter             427               248087   
9        Pena High School     Charter             962               585858   
10     Wright High School     Charter            1800              1049400   
11  Rodriguez High School    District            3999              2547363   
12    Johnson High School    District            4761              3094650   
13       Ford High School    District            2739              1763916   
14     Thomas High School     Charter            1635              1043130   

    Per Student Budget  Aver. Reading Score  Aver. Math Score  
0                655.0                81.18             76.63  
1                639.0                81.16             76.71  
2                600.0                83.73             83.36  
3                652.0                80.93             77.29  
4                625.0                83.82             83.35  
5                578.0                83.99             83.27  
6                582.0                83.98             83.06  
7                628.0                81.03             77.05  
8                581.0                83.81             83.80  
9                609.0                84.04             83.84  
10               583.0                83.96             83.68  
11               637.0                80.74             76.84  
12               650.0                80.97             77.07  
13               644.0                80.75             77.10  
14               638.0                83.85             83.42

In [72]:
#Counting how many students passed reading for each school

count_passing = merge_df.loc[:,['school_name','reading_score']]
count_passing['reading_score']=count_passing.loc[count_passing['reading_score']>=70]
count_passing=count_passing.groupby('school_name')
count_passing=pd.DataFrame(count_passing.count())
count_passing


reading_score
school_name                         
Bailey High School              4077
Cabrera High School             1803
Figueroa High School            2381
Ford High School                2172
Griffin High School             1426
Hernandez High School           3748
Holden High School               411
Huang High School               2372
Johnson High School             3867
Pena High School                 923
Rodriguez High School           3208
Shelton High School             1688
Thomas High School              1591
Wilson High School              2204
Wright High School              1739

In [73]:
#Counting number of students for each school
count_student = merge_df.loc[:,['school_name','size']]
count_student=count_student.groupby('school_name')
count_student=pd.DataFrame(count_student.count())
count_student

size
school_name                
Bailey High School     4976
Cabrera High School    1858
Figueroa High School   2949
Ford High School       2739
Griffin High School    1468
Hernandez High School  4635
Holden High School      427
Huang High School      2917
Johnson High School    4761
Pena High School        962
Rodriguez High School  3999
Shelton High School    1761
Thomas High School     1635
Wilson High School     2283
Wright High School     1800

In [74]:
#Counting how many students passed math for each school

count_passing_math = merge_df.loc[:,['school_name','math_score']]
count_passing_math['math_score']=count_passing_math.loc[count_passing_math['math_score']>=70]
count_passing_math=count_passing_math.groupby('school_name')
count_passing_math=pd.DataFrame(count_passing_math.count())
count_passing_math

math_score
school_name                      
Bailey High School           3318
Cabrera High School          1749
Figueroa High School         1946
Ford High School             1871
Griffin High School          1371
Hernandez High School        3094
Holden High School            395
Huang High School            1916
Johnson High School          3145
Pena High School              910
Rodriguez High School        2654
Shelton High School          1653
Thomas High School           1525
Wilson High School           2143
Wright High School           1680

In [75]:
#Joining counts for math, reading>=70 and total students' count
joined_counts=pd.merge(count_passing_math, count_passing,
how='left', on='school_name')
joined_counts1=pd.merge(joined_counts,count_student, how='left',on='school_name')
joined_counts1=joined_counts1.reset_index()
joined_counts1.rename(columns={'school_name':'School Name',
                               'math_score':'Total Count Passing Math',
                               'reading_score':'Total Count Passing Reading',
                               'size':'Total Students'},inplace=True)
joined_counts1

School Name  Total Count Passing Math  \
0      Bailey High School                      3318   
1     Cabrera High School                      1749   
2    Figueroa High School                      1946   
3        Ford High School                      1871   
4     Griffin High School                      1371   
5   Hernandez High School                      3094   
6      Holden High School                       395   
7       Huang High School                      1916   
8     Johnson High School                      3145   
9        Pena High School                       910   
10  Rodriguez High School                      2654   
11    Shelton High School                      1653   
12     Thomas High School                      1525   
13     Wilson High School                      2143   
14     Wright High School                      1680   

    Total Count Passing Reading  Total Students  
0                          4077            4976  
1                          1803            1858  
2                          2381            2949  
3                          2172            2739  
4                          1426            1468  
5                          3748            4635  
6                           411             427  
7                          2372            2917  
8                          3867            4761  
9                           923             962  
10                         3208            3999  
11                         1688            1761  
12                         1591            1635  
13                         2204            2283  
14                         1739            1800

In [76]:
#Figuring out counts of students that passed both math and reading
count_both=merge_df.loc[:,['school_name','reading_score','math_score']]
count_both2=count_both[(count_both['reading_score']>=70) & (count_both['math_score']>=70)]
count_both2=count_both2.groupby('school_name')
count_both2=pd.DataFrame(count_both2.count())

count_both3=count_both2.drop(['reading_score'], axis=1)
count_both3=count_both3.reset_index()
count_both3.rename(columns={'school_name':'School Name',
                               'math_score':'Overall Count',
                               },inplace=True)

count_both3

School Name  Overall Count
0      Bailey High School           2719
1     Cabrera High School           1697
2    Figueroa High School           1569
3        Ford High School           1487
4     Griffin High School           1330
5   Hernandez High School           2481
6      Holden High School            381
7       Huang High School           1561
8     Johnson High School           2549
9        Pena High School            871
10  Rodriguez High School           2119
11    Shelton High School           1583
12     Thomas High School           1487
13     Wilson High School           2068
14     Wright High School           1626

In [77]:
#Merging joint counts and overall counts
joined_df=pd.merge(count_both3,joined_counts1, how='left',on='School Name')
joined_df


School Name  Overall Count  Total Count Passing Math  \
0      Bailey High School           2719                      3318   
1     Cabrera High School           1697                      1749   
2    Figueroa High School           1569                      1946   
3        Ford High School           1487                      1871   
4     Griffin High School           1330                      1371   
5   Hernandez High School           2481                      3094   
6      Holden High School            381                       395   
7       Huang High School           1561                      1916   
8     Johnson High School           2549                      3145   
9        Pena High School            871                       910   
10  Rodriguez High School           2119                      2654   
11    Shelton High School           1583                      1653   
12     Thomas High School           1487                      1525   
13     Wilson High School           2068                      2143   
14     Wright High School           1626                      1680   

    Total Count Passing Reading  Total Students  
0                          4077            4976  
1                          1803            1858  
2                          2381            2949  
3                          2172            2739  
4                          1426            1468  
5                          3748            4635  
6                           411             427  
7                          2372            2917  
8                          3867            4761  
9                           923             962  
10                         3208            3999  
11                         1688            1761  
12                         1591            1635  
13                         2204            2283  
14                         1739            1800

In [78]:
joined_df["% Overall Passing"]=round(joined_df['Overall Count']/joined_df['Total Students']*100,2)
joined_df["% Passing Math"]=round(joined_df['Total Count Passing Math']/joined_df['Total Students']*100,2)
joined_df["% Passing Reading"]=round(joined_df['Total Count Passing Reading']/joined_df['Total Students']*100,2)
joined_df

School Name  Overall Count  Total Count Passing Math  \
0      Bailey High School           2719                      3318   
1     Cabrera High School           1697                      1749   
2    Figueroa High School           1569                      1946   
3        Ford High School           1487                      1871   
4     Griffin High School           1330                      1371   
5   Hernandez High School           2481                      3094   
6      Holden High School            381                       395   
7       Huang High School           1561                      1916   
8     Johnson High School           2549                      3145   
9        Pena High School            871                       910   
10  Rodriguez High School           2119                      2654   
11    Shelton High School           1583                      1653   
12     Thomas High School           1487                      1525   
13     Wilson High School           2068                      2143   
14     Wright High School           1626                      1680   

    Total Count Passing Reading  Total Students  % Overall Passing  \
0                          4077            4976              54.64   
1                          1803            1858              91.33   
2                          2381            2949              53.20   
3                          2172            2739              54.29   
4                          1426            1468              90.60   
5                          3748            4635              53.53   
6                           411             427              89.23   
7                          2372            2917              53.51   
8                          3867            4761              53.54   
9                           923             962              90.54   
10                         3208            3999              52.99   
11                         1688            1761              89.89   
12                         1591            1635              90.95   
13                         2204            2283              90.58   
14                         1739            1800              90.33   

    % Passing Math  % Passing Reading  
0            66.68              81.93  
1            94.13              97.04  
2            65.99              80.74  
3            68.31              79.30  
4            93.39              97.14  
5            66.75              80.86  
6            92.51              96.25  
7            65.68              81.32  
8            66.06              81.22  
9            94.59              95.95  
10           66.37              80.22  
11           93.87              95.85  
12           93.27              97.31  
13           93.87              96.54  
14           93.33              96.61

In [79]:
#Joining school summary and tests summary tables together
school_summary_df=pd.merge(add,joined_df, how='left',on=['School Name','Total Students'])
school_summary_df=school_summary_df.drop(['Overall Count', 'Total Count Passing Math','Total Count Passing Reading'], axis=1)
school_summary_df=school_summary_df[['School Name',
                                     'School Type',
                                     'Total Students',
                                     'Total School Budget',
                                     'Per Student Budget',
                                     'Aver. Math Score',
                                     'Aver. Reading Score',
                                     '% Passing Math',
                                     '% Passing Reading',
                                      '% Overall Passing']]
school_summary_df

School Name School Type  Total Students  Total School Budget  \
0       Huang High School    District            2917              1910635   
1    Figueroa High School    District            2949              1884411   
2     Shelton High School     Charter            1761              1056600   
3   Hernandez High School    District            4635              3022020   
4     Griffin High School     Charter            1468               917500   
5      Wilson High School     Charter            2283              1319574   
6     Cabrera High School     Charter            1858              1081356   
7      Bailey High School    District            4976              3124928   
8      Holden High School     Charter             427               248087   
9        Pena High School     Charter             962               585858   
10     Wright High School     Charter            1800              1049400   
11  Rodriguez High School    District            3999              2547363   
12    Johnson High School    District            4761              3094650   
13       Ford High School    District            2739              1763916   
14     Thomas High School     Charter            1635              1043130   

    Per Student Budget  Aver. Math Score  Aver. Reading Score  % Passing Math  \
0                655.0             76.63                81.18           65.68   
1                639.0             76.71                81.16           65.99   
2                600.0             83.36                83.73           93.87   
3                652.0             77.29                80.93           66.75   
4                625.0             83.35                83.82           93.39   
5                578.0             83.27                83.99           93.87   
6                582.0             83.06                83.98           94.13   
7                628.0             77.05                81.03           66.68   
8                581.0             83.80                83.81           92.51   
9                609.0             83.84                84.04           94.59   
10               583.0             83.68                83.96           93.33   
11               637.0             76.84                80.74           66.37   
12               650.0             77.07                80.97           66.06   
13               644.0             77.10                80.75           68.31   
14               638.0             83.42                83.85           93.27   

    % Passing Reading  % Overall Passing  
0               81.32              53.51  
1               80.74              53.20  
2               95.85              89.89  
3               80.86              53.53  
4               97.14              90.60  
5               96.54              90.58  
6               97.04              91.33  
7               81.93              54.64  
8               96.25              89.23  
9               95.95              90.54  
10              96.61              90.33  
11              80.22              52.99  
12              81.22              53.54  
13              79.30              54.29  
14              97.31              90.95

# Top Performing Schools (By % Overall Passing)
Create a table that highlights the top 5 performing schools based on % Overall Passing. Include:
School Name
School Type
Total Students
Total School Budget
Per Student Budget
Average Math Score
Average Reading Score
% Passing Math (The percentage of students that passed math.)
% Passing Reading (The percentage of students that passed reading.)
% Overall Passing (The percentage of students that passed math and reading.)

In [80]:
#Sorting top 5 performing schools based on overall passing
top_5 = school_summary_df.sort_values("% Overall Passing", ascending=False)
top_5=top_5.set_index('School Name')
top_5.head(5)

School Type  Total Students  Total School Budget  \
School Name                                                            
Cabrera High School     Charter            1858              1081356   
Thomas High School      Charter            1635              1043130   
Griffin High School     Charter            1468               917500   
Wilson High School      Charter            2283              1319574   
Pena High School        Charter             962               585858   

                     Per Student Budget  Aver. Math Score  \
School Name                                                 
Cabrera High School               582.0             83.06   
Thomas High School                638.0             83.42   
Griffin High School               625.0             83.35   
Wilson High School                578.0             83.27   
Pena High School                  609.0             83.84   

                     Aver. Reading Score  % Passing Math  % Passing Reading  \
School Name                                                                   
Cabrera High School                83.98           94.13              97.04   
Thomas High School                 83.85           93.27              97.31   
Griffin High School                83.82           93.39              97.14   
Wilson High School                 83.99           93.87              96.54   
Pena High School                   84.04           94.59              95.95   

                     % Overall Passing  
School Name                             
Cabrera High School              91.33  
Thomas High School               90.95  
Griffin High School              90.60  
Wilson High School               90.58  
Pena High School                 90.54

# Bottom Performing Schools (By % Overall Passing)
Create a table that highlights the bottom 5 performing schools based on % Overall Passing. Include all of the same metrics as above.


In [81]:
#Sorting bottom 5 performing schools based on overall passing
worst_5 = school_summary_df.sort_values("% Overall Passing", ascending=True)
worst_5=worst_5.set_index('School Name')
worst_5.head(5)


School Type  Total Students  Total School Budget  \
School Name                                                              
Rodriguez High School    District            3999              2547363   
Figueroa High School     District            2949              1884411   
Huang High School        District            2917              1910635   
Hernandez High School    District            4635              3022020   
Johnson High School      District            4761              3094650   

                       Per Student Budget  Aver. Math Score  \
School Name                                                   
Rodriguez High School               637.0             76.84   
Figueroa High School                639.0             76.71   
Huang High School                   655.0             76.63   
Hernandez High School               652.0             77.29   
Johnson High School                 650.0             77.07   

                       Aver. Reading Score  % Passing Math  % Passing Reading  \
School Name                                                                     
Rodriguez High School                80.74           66.37              80.22   
Figueroa High School                 81.16           65.99              80.74   
Huang High School                    81.18           65.68              81.32   
Hernandez High School                80.93           66.75              80.86   
Johnson High School                  80.97           66.06              81.22   

                       % Overall Passing  
School Name                               
Rodriguez High School              52.99  
Figueroa High School               53.20  
Huang High School                  53.51  
Hernandez High School              53.53  
Johnson High School                53.54

# Math Scores by Grade
Create a table that lists the average Math Score for students of each grade level (9th, 10th, 11th, 12th) at each school.



In [82]:
d3 = merge_df.groupby(['school_name','grade'])
score1 = pd.DataFrame(round(d3.mean(),2))    
score1 = score1.loc[:,['math_score']]
score1.head()


score1=score1.reset_index()

In [83]:
pivoted = score1.pivot(index='school_name', columns='grade', values='math_score')
pivoted.columns.name=None
pivoted.columns=['9th','10th', '11th', '12th']
pivoted.reset_index()

school_name    9th   10th   11th   12th
0      Bailey High School  77.00  77.52  76.49  77.08
1     Cabrera High School  83.15  82.77  83.28  83.09
2    Figueroa High School  76.54  76.88  77.15  76.40
3        Ford High School  77.67  76.92  76.18  77.36
4     Griffin High School  84.23  83.84  83.36  82.04
5   Hernandez High School  77.34  77.14  77.19  77.44
6      Holden High School  83.43  85.00  82.86  83.79
7       Huang High School  75.91  76.45  77.23  77.03
8     Johnson High School  76.69  77.49  76.86  77.19
9        Pena High School  83.37  84.33  84.12  83.63
10  Rodriguez High School  76.61  76.40  77.69  76.86
11    Shelton High School  82.92  83.38  83.78  83.42
12     Thomas High School  83.09  83.50  83.50  83.59
13     Wilson High School  83.72  83.20  83.04  83.09
14     Wright High School  84.01  83.84  83.64  83.26

# Reading Scores by Grade
Create a table that lists the average Reading Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

In [84]:
d4 = merge_df.groupby(['school_name','grade'])
score2 = pd.DataFrame(round(d4.mean(),2))    
score2 = score2.loc[:,['reading_score']]
score2.head()

score2=score2.reset_index()

pivoted_reading = score2.pivot(index='school_name', columns='grade', values='reading_score')
pivoted_reading.columns.name=None
pivoted_reading.columns=['9th','10th', '11th', '12th']
pivoted_reading.reset_index()

school_name    9th   10th   11th   12th
0      Bailey High School  80.91  80.95  80.91  81.30
1     Cabrera High School  84.25  83.79  84.29  83.68
2    Figueroa High School  81.41  80.64  81.38  81.20
3        Ford High School  81.26  80.40  80.66  80.63
4     Griffin High School  83.71  84.29  84.01  83.37
5   Hernandez High School  80.66  81.40  80.86  80.87
6      Holden High School  83.32  83.82  84.70  83.68
7       Huang High School  81.51  81.42  80.31  81.29
8     Johnson High School  80.77  80.62  81.23  81.26
9        Pena High School  83.61  84.34  84.59  83.81
10  Rodriguez High School  80.63  80.86  80.38  80.99
11    Shelton High School  83.44  84.37  82.78  84.12
12     Thomas High School  84.25  83.59  83.83  83.73
13     Wilson High School  84.02  83.76  84.32  83.94
14     Wright High School  83.81  84.16  84.07  83.83

# Scores by School Spending
Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
Average Math Score
Average Reading Score
% Passing Math (The percentage of students that passed math.)
% Passing Reading (The percentage of students that passed reading.)
% Overall Passing (The percentage of students that passed math and reading.)

In [85]:
#Figuring out max and min school budget to create bins
print(schools["Per Student Budget"].max())
print(schools["Per Student Budget"].min())

655.0
578.0


In [86]:
# Creating bins in which to place values based on per student budget(spending)
bins = [0, 585, 630, 645, 675]

# Create labels for these bins
schools2=school_summary_df.copy()
group_labels = ["< 585", "585 - 630", "630 - 645", "645 - 675"]
schools2['Spending Ranges Per Student']=pd.cut(schools["Per Student Budget"], bins, labels=group_labels)
schools2


School Name School Type  Total Students  Total School Budget  \
0       Huang High School    District            2917              1910635   
1    Figueroa High School    District            2949              1884411   
2     Shelton High School     Charter            1761              1056600   
3   Hernandez High School    District            4635              3022020   
4     Griffin High School     Charter            1468               917500   
5      Wilson High School     Charter            2283              1319574   
6     Cabrera High School     Charter            1858              1081356   
7      Bailey High School    District            4976              3124928   
8      Holden High School     Charter             427               248087   
9        Pena High School     Charter             962               585858   
10     Wright High School     Charter            1800              1049400   
11  Rodriguez High School    District            3999              2547363   
12    Johnson High School    District            4761              3094650   
13       Ford High School    District            2739              1763916   
14     Thomas High School     Charter            1635              1043130   

    Per Student Budget  Aver. Math Score  Aver. Reading Score  % Passing Math  \
0                655.0             76.63                81.18           65.68   
1                639.0             76.71                81.16           65.99   
2                600.0             83.36                83.73           93.87   
3                652.0             77.29                80.93           66.75   
4                625.0             83.35                83.82           93.39   
5                578.0             83.27                83.99           93.87   
6                582.0             83.06                83.98           94.13   
7                628.0             77.05                81.03           66.68   
8                581.0             83.80                83.81           92.51   
9                609.0             83.84                84.04           94.59   
10               583.0             83.68                83.96           93.33   
11               637.0             76.84                80.74           66.37   
12               650.0             77.07                80.97           66.06   
13               644.0             77.10                80.75           68.31   
14               638.0             83.42                83.85           93.27   

    % Passing Reading  % Overall Passing Spending Ranges Per Student  
0               81.32              53.51                   645 - 675  
1               80.74              53.20                   630 - 645  
2               95.85              89.89                   585 - 630  
3               80.86              53.53                   645 - 675  
4               97.14              90.60                   585 - 630  
5               96.54              90.58                       < 585  
6               97.04              91.33                       < 585  
7               81.93              54.64                   585 - 630  
8               96.25              89.23                       < 585  
9               95.95              90.54                   585 - 630  
10              96.61              90.33                       < 585  
11              80.22              52.99                   630 - 645  
12              81.22              53.54                   645 - 675  
13              79.30              54.29                   630 - 645  
14              97.31              90.95                   630 - 645

In [97]:
summary_by_spending=schools2.groupby(by="Spending Ranges Per Student").mean()
summary_by_spending=summary_by_spending.drop(['Total Students','Total School Budget','Per Student Budget'],axis=1)
summary_by_spending.round({'Aver. Math Score':2,'Aver. Reading Score':2,'% Passing Math':2,'% Passing Reading':2,"% Overall Passing":2})

Aver. Math Score  Aver. Reading Score  \
Spending Ranges Per Student                                          
< 585                                   83.45                83.94   
585 - 630                               81.90                83.16   
630 - 645                               78.52                81.62   
645 - 675                               77.00                81.03   

                             % Passing Math  % Passing Reading  \
Spending Ranges Per Student                                      
< 585                                 93.46              96.61   
585 - 630                             87.13              92.72   
630 - 645                             73.48              84.39   
645 - 675                             66.16              81.13   

                             % Overall Passing  
Spending Ranges Per Student                     
< 585                                    90.37  
585 - 630                                81.42  
630 - 645                                62.86  
645 - 675                                53.53

# Scores by School Size
Repeat the above breakdown, but this time group schools based on a reasonable approximation of school size (Small, Medium, Large).

In [91]:
#Figuring out max and min school size to create bins
print(schools["Total Students"].max())
print(schools["Total Students"].min())

4976
427


In [92]:
schools3=school_summary_df.copy()
#Creating bins

size_bins = [0, 1000, 2000, 5000]

# Create labels for these bins
size_labels = ["Small (< 1000)", "Medium (1000-2000)", "Large (2000-5000)"]
schools3["School Size"]=pd.cut(school_summary_df["Total Students"], size_bins, labels=size_labels)
schools3


School Name School Type  Total Students  Total School Budget  \
0       Huang High School    District            2917              1910635   
1    Figueroa High School    District            2949              1884411   
2     Shelton High School     Charter            1761              1056600   
3   Hernandez High School    District            4635              3022020   
4     Griffin High School     Charter            1468               917500   
5      Wilson High School     Charter            2283              1319574   
6     Cabrera High School     Charter            1858              1081356   
7      Bailey High School    District            4976              3124928   
8      Holden High School     Charter             427               248087   
9        Pena High School     Charter             962               585858   
10     Wright High School     Charter            1800              1049400   
11  Rodriguez High School    District            3999              2547363   
12    Johnson High School    District            4761              3094650   
13       Ford High School    District            2739              1763916   
14     Thomas High School     Charter            1635              1043130   

    Per Student Budget  Aver. Math Score  Aver. Reading Score  % Passing Math  \
0                655.0             76.63                81.18           65.68   
1                639.0             76.71                81.16           65.99   
2                600.0             83.36                83.73           93.87   
3                652.0             77.29                80.93           66.75   
4                625.0             83.35                83.82           93.39   
5                578.0             83.27                83.99           93.87   
6                582.0             83.06                83.98           94.13   
7                628.0             77.05                81.03           66.68   
8                581.0             83.80                83.81           92.51   
9                609.0             83.84                84.04           94.59   
10               583.0             83.68                83.96           93.33   
11               637.0             76.84                80.74           66.37   
12               650.0             77.07                80.97           66.06   
13               644.0             77.10                80.75           68.31   
14               638.0             83.42                83.85           93.27   

    % Passing Reading  % Overall Passing         School Size  
0               81.32              53.51   Large (2000-5000)  
1               80.74              53.20   Large (2000-5000)  
2               95.85              89.89  Medium (1000-2000)  
3               80.86              53.53   Large (2000-5000)  
4               97.14              90.60  Medium (1000-2000)  
5               96.54              90.58   Large (2000-5000)  
6               97.04              91.33  Medium (1000-2000)  
7               81.93              54.64   Large (2000-5000)  
8               96.25              89.23      Small (< 1000)  
9               95.95              90.54      Small (< 1000)  
10              96.61              90.33  Medium (1000-2000)  
11              80.22              52.99   Large (2000-5000)  
12              81.22              53.54   Large (2000-5000)  
13              79.30              54.29   Large (2000-5000)  
14              97.31              90.95  Medium (1000-2000)

In [98]:
summary_by_size=schools3.groupby(by="School Size").mean()

summary_by_size=summary_by_size.round({'Aver. Math Score':2,'Aver. Reading Score':2,'% Passing Math':2,'% Passing Reading':2,"% Overall Passing":2})
summary_by_size.drop(['Total Students','Total School Budget','Per Student Budget'],axis=1)

Aver. Math Score  Aver. Reading Score  % Passing Math  \
School Size                                                                 
Small (< 1000)                 83.82                83.93           93.55   
Medium (1000-2000)             83.37                83.87           93.60   
Large (2000-5000)              77.74                81.34           69.96   

                    % Passing Reading  % Overall Passing  
School Size                                               
Small (< 1000)                  96.10              89.88  
Medium (1000-2000)              96.79              90.62  
Large (2000-5000)               82.77              58.28

# Scores by School Type


In [99]:
#Calculating test scores averages by school type
schools4=school_summary_df.copy()
school_by_type=schools4.groupby(by="School Type").mean()
school_by_type=school_by_type.drop(columns=['Total Students','Total School Budget','Per Student Budget'])
school_by_type.round({'Aver. Math Score':2,'Aver. Reading Score':2,'% Passing Math':2,'% Passing Reading':2,"% Overall Passing":2})





Aver. Math Score  Aver. Reading Score  % Passing Math  \
School Type                                                          
Charter                 83.47                83.90           93.62   
District                76.96                80.97           66.55   

             % Passing Reading  % Overall Passing  
School Type                                        
Charter                  96.59              90.43  
District                 80.80              53.67

# Analysis



There are 15 schools, of which 7 are district and 8 are charter schools. According to the school summary table, charter schools receive less school budget compared to district schools, yet their overall passing percentages are higher than district schools. For example, the top 5 schools are all charter schools and bottom 5 schools are district schools. If we look at school size, all the district schools are in "large school" category with 2000-5000 students, that has overall passing percentage of 58.28%. Small and medium-sized schools are all charter schools with maximum number of students being 2000 and overall percentages are about 89.88& and 90.62%. If we look into test scores based on school type, charter schools have way high overall passing percentage of 90.43 compared to district schools' overall passing being 53.67%. These data shows that in this group of 15 schools, charter schools are doing way better despite of having less school budget. This might be due to smaller and medium classroom sizes, where students get more teacher time and individualized attention and instruction. 
